File: 04-structural-patterns/02-adapter/02-adapter.go
package main

import "fmt"

// Target interface
type PaymentProcessor interface {
	Pay(amount float64)
}

// Adaptees
type PayPalService struct{}

func (p *PayPalService) SendPayment(amount float64) {
	fmt.Printf("PayPal processed payment of %.2f\n", amount)
}

type StripeService struct{}

func (s *StripeService) ChargeCard(amount float64) {
	fmt.Printf("Stripe charged %.2f\n", amount)
}

type RazorpayService struct{}

func (r *RazorpayService) MakeTransaction(amount float64) {
	fmt.Printf("Razorpay transaction of %.2f\n", amount)
}

// Adapters
type PayPalAdapter struct{ service *PayPalService }

func (a *PayPalAdapter) Pay(amount float64) {
	a.service.SendPayment(amount)
}

type StripeAdapter struct{ service *StripeService }

func (a *StripeAdapter) Pay(amount float64) {
	a.service.ChargeCard(amount)
}

type RazorpayAdapter struct{ service *RazorpayService }

func (a *RazorpayAdapter) Pay(amount float64) {
	a.service.MakeTransaction(amount)
}

// Client
func processPayment(p PaymentProcessor, amount float64) {
	p.Pay(amount)
}

func main() {
	payments := []PaymentProcessor{
		&PayPalAdapter{&PayPalService{}},
		&StripeAdapter{&StripeService{}},
		&RazorpayAdapter{&RazorpayService{}},
	}

	for _, p := range payments {
		processPayment(p, 1000)
	}
}

File: 04-structural-patterns/02-adapter/02-adapter.py
# Target interface
class PaymentProcessor:
    def pay(self, amount):
        raise NotImplementedError()

# Adaptees (Third-party services)
class PayPalService:
    def send_payment(self, amount):
        print(f"PayPal processed payment of {amount}")

class StripeService:
    def charge_card(self, amount):
        print(f"Stripe charged amount {amount}")

class RazorpayService:
    def make_transaction(self, amount):
        print(f"Razorpay processed transaction of {amount}")

# Adapters
class PayPalAdapter(PaymentProcessor):
    def __init__(self, service):
        self.service = service

    def pay(self, amount):
        self.service.send_payment(amount)

class StripeAdapter(PaymentProcessor):
    def __init__(self, service):
        self.service = service

    def pay(self, amount):
        self.service.charge_card(amount)

class RazorpayAdapter(PaymentProcessor):
    def __init__(self, service):
        self.service = service

    def pay(self, amount):
        self.service.make_transaction(amount)

# Client code
def process_payment(processor: PaymentProcessor, amount: int):
    processor.pay(amount)

# Usage
gateways = [
    PayPalAdapter(PayPalService()),
    StripeAdapter(StripeService()),
    RazorpayAdapter(RazorpayService())
]

for gateway in gateways:
    process_payment(gateway, 1000)
